### DISCUSSION TITLE
Why pydantic model validator in async path operation function run in main thread?

### DISCUSSION THREAD

**Discussion by BaekjunKim2** [fastapi/fastapi#9462] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
#fastapi/routing.py
async def serialize_response(
    *,
    field: Optional[ModelField] = None,
    response_content: Any,
    include: Optional[Union[SetIntStr, DictIntStrAny]] = None,
    exclude: Optional[Union[SetIntStr, DictIntStrAny]] = None,
    by_alias: bool = True,
    exclude_unset: bool = False,
    exclude_defaults: bool = False,
    exclude_none: bool = False,
    is_coroutine: bool = True,
) -> Any:
    if field:
        errors = []
        response_content = _prepare_response_content(
            response_content,
            exclude_unset=exclude_unset,
            exclude_defaults=exclude_defaults,
            exclude_none=exclude_none,
        )
        if is_coroutine:
            value, errors_ = field.validate(response_content, {}, loc=("response",))
        else:
            value, errors_ = await run_in_threadpool(
                field.validate, response_content, {}, loc=("response",)
            )
...
```

### Description

I'm developing a web application using FastAPI and trying to use async path operation function as much as possible to become familiar with it.
However, I have discovered that pydantic's validator cannot be async, as explained in this issue on it: https://github.com/pydantic/pydantic/issues/857

In the `serialize_response()` function above, the field validating is called in main thread's event loop if the path operation function is async. (is_coroutine is True)
This concerns me because when pydantic validator takes long time, it may block the event loop.
I wonder why the validator is not called in a thread pool instead.

### Operating System

Linux

### Operating System Details

Linux 5.15.90.1-microsoft-standard-WSL2 #1 SMP Fri Jan 27 02:56:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### FastAPI Version

0.95.1

### Python Version

3.10.7

**Comment by adriangb** [comment]
Pydantic V1 is a pure python (well, cythonized but still) package. So you'd have to run in processes, which is super heavyweight. Pydantic V2 is written in Rust. It doesn't currently release the GIL _but it probably could, in some scenarios_. It's not an immediate priority for the Pydantic team, _but there is hope_.

**Reply by davidbrochart** [reply]
> Pydantic V1 is a pure python (well, cythonized but still) package. So you'd have to run in processes, which is super heavyweight.

I don't understand, if it were pure python (forgetting about cython), it could run in a thread and we could await that, it wouldn't block the event loop.

> Pydantic V2 is written in Rust. It doesn't currently release the GIL _but it probably could, in some scenarios_. It's not an immediate priority for the Pydantic team, _but there is hope_.

Curious to know if things have changed?

**Comment by methane** [comment, marked as answer]
https://github.com/tiangolo/fastapi/pull/888 is the PR that starts using threadpool for validation.

It expect if path operation function is async, returned object doesn't contain blocking operation during serialization.

I don't like current behavior because I return only pure pydantic model object without blocking operations. Running in a threadpool has significant overhead.
